oss-internship-2020/sapi_lodepng/examples/main_sandboxed.cc

```diff
@@ -0,0 +1,205 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <filesystem>
+#include <iostream>
+
+#include "sandbox.h"
+#include "sandboxed_api/util/flag.h"
+
+ABSL_FLAG(string, images_path, std::filesystem::current_path().string(),
+          "path to the folder containing test images");
+
+void generate_one_step(SapiLodepngSandbox &sandbox, LodepngApi &api) {
+  constexpr unsigned int width = 512, height = 512,
+                         img_len = width * height * 4;
+  std::vector<unsigned char> image(img_len);
+
+  for (int y = 0; y < height; ++y) {
+    for (int x = 0; x < width; ++x) {
+      image[4 * width * y + 4 * x + 0] = 255 * !(x & y);
+      image[4 * width * y + 4 * x + 1] = x ^ y;
+      image[4 * width * y + 4 * x + 2] = x | y;
+      image[4 * width * y + 4 * x + 3] = 255;
+    }
+  }
+
+  // encode the image
+  sapi::v::Array<unsigned char> sapi_image(image.data(), img_len);
+  sapi::v::UInt sapi_width(width), sapi_height(height);
+  sapi::v::ConstCStr sapi_filename("/output/out_generated1.png");
+
+  sapi::StatusOr<unsigned int> result = api.lodepng_encode32_file(
+      sapi_filename.PtrBefore(), sapi_image.PtrBefore(), sapi_width.GetValue(),
+      sapi_height.GetValue());
+
+  assert(result.ok());
+  assert(!result.value());
+
+  // after the image has been encoded, decode it to check that the
+  // pixel values are the same
+
+  sapi::v::UInt sapi_width2, sapi_height2;
+  sapi::v::IntBase<unsigned char *> sapi_image_ptr(0);
+
+  result = api.lodepng_decode32_file(
+      sapi_image_ptr.PtrBoth(), sapi_width2.PtrBoth(), sapi_height2.PtrBoth(),
+      sapi_filename.PtrBefore());
+  assert(result.ok());
+  assert(!result.value());
+
+  assert(sapi_width2.GetValue() == width);
+  assert(sapi_height2.GetValue() == height);
+
+  // the pixels have been allocated inside the sandboxed process
+  // memory, so we need to transfer them to this process.
+  // Transferring the memory has the following steps:
+  // 1) define a RemotePtr variable that holds the memory location from
+  // the sandboxed process
+  // 2) define an array with the required length
+  // 3) set the remote pointer for the array to specify where the memory
+  // that will be transferred is located
+  // 4) transfer the memory to this process (this step is why we need
+  // the pointer and the length)
+  sapi::v::RemotePtr sapi_remote_out_ptr(sapi_image_ptr.GetValue());
```

Comment:
no need for the `sapi::v::RemotePtr` just use `sapi_image_ptr.GetValue()` in `SetRemote` call below.